### PR TITLE
optimization of Allpairs in the symmetric case

### LINKED ISSUE #731: optimization of Allpairs in the symmetric case
Currently, `Allpairs(setA,setB, Func)` applies `Func` on each pair of A[i] and B[j]. If the length of A is m, and the length of B is n, `Func` will execute `m x n` times totally.

Problem:
In the case of A=B (i.e., m=n, A[i] = B[i]), `Func(A[i], B[j])` and `Func(A[j], B[i])` are duplicate. Totally nearly half of the executions of Func are unnecessary. 
We need to optimize Allpairs to avoid doing the same computing twice.

Solution:
If i<=j, do Func(A[i], B[j]); if i>j, do nothing.

Potential Problem:
`allpairs_master` can set the width and height of one work unit. The modification of `allpairs_master` should try to avoid splitting work units unevenly.

### PR DESCRIPTION
**Pull request by hmeng-19**
add --symmetric option to allparis_multicore and allpairs_master
	this option allows allpairs compute only half of a symmetric matrix

Add --index option into allpairs_multicore
	this option allows the user to specify the indexes of a matrix

This branch tries to fix #731

### CODE REVIEW

**Comment by dthain** [comment]
Looks pretty good, except one more piece needed:

In the master, skip generating a sub-task if it is completely on the "wrong" side of the diagonal.  As written, the sub task will be generated and do no work (which is ok) but it will still result in sending all the input files to the empty task.

**Comment by hmeng-19** [comment]
Good point.

**Comment by hmeng-19** [comment]
@dthain , I added a new commit to fix the problem you pointed out. 
Please recheck it.

**Comment by dthain** [comment]
Overall, looks good.
You could simplify the option parsing by using something like `n=sscanf(optarg,"%d,%d",&x,&y)` in main.

**Comment by hmeng-19** [comment]
The only drawback of using `sscanf` is that I have no way to judge whether the option is valid or not. If the user set --index to be `3 6 8`, I want to tell him that his option is wrong.

**Comment by hmeng-19** [comment]
BTW, `string_split` may be a little more expensive. But I think it is a good fit here. We only call it once.

**Comment by dthain** [comment]
sscanf returns the number of fields matched, so that you can check.  Performance is not my concern, I am simply showing you how to make your C code even more compact and clear.

**Comment by hmeng-19** [comment]
You are right. I even forgot the return value of `sscanf` (shameful...). It seems I really need to take your compiler class to review C.

**Comment by hmeng-19** [comment]
@dthain , it is there. Hope no more bug. :)

**Comment by dthain** [comment]
Beautiful.  :)  Merge it.

File: allpairs/src/allpairs_master.c [hunk 1]
```diff
@@ -64,6 +71,7 @@ static void show_help(const char *cmd)
 	fprintf(stdout, " %-30s Set the project name to <project>\n", "-N,--project-name=<project>");
 	fprintf(stdout, " %-30s Priority. Higher the value, higher the priority.\n", "-P,--priority=<integer>");
 	fprintf(stdout, " %-30s Enable debugging for this subsystem.  (Try -d all to start.)\n", "-d,--debug=<flag>");
+	fprintf(stdout, " %-30s Compute half of a symmetric matrix.\n", "   --symmetric");
```
Comment (btovar) [inline allpairs/src/allpairs_master.c]: Is it a matrix? 
If not, I rather have --commutative, to mean that F(a,b) = F(b,a).

**Comment by hmeng-19** [comment]
@btovar , this is a symmetric matrix and F(a, b) = F(b, a).
I think `--symmetric` is a good fit here.
Thanks for your suggestion. :)

**Pull request merged**